### PR TITLE
Move cargo fmt checks and some other jobs to stable, ubuntu-latest

### PR DESCRIPTION
**Pull request by Veetaha**
## Description

See https://github.com/amethyst/amethyst/pull/2443#issuecomment-677934141

## Modifications

- Now a bunch of jobs that were previously selected to be run on `macos` switched to `ubuntu-latest`, cargo fmt check is one of them

## PR Checklist

By placing an x in the boxes I certify that I have:

- [ ] Added unit tests for new code added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.
- [ ] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Updated the content of the book if this PR would make the book outdated.

If this modified or created any rs files:

- [ ] Ran `cargo +stable fmt --all`
- [ ] Ran `cargo clippy --workspace --features "empty"` (may require `cargo clean` before)
- [ ] Ran `cargo build --features "empty"`
- [ ] Ran `cargo test --workspace --features "empty"`

### CODE REVIEW

**Review by CleanCut — CHANGES_REQUESTED** [review]
Caught a couple things you missed on the mdbook-linkcheck section. I know it's commented out, but let's keep it in good shape for when we uncomment it (hopefully really soon!). 😄 

#

I find the runtimes of [this CI run](https://github.com/amethyst/amethyst/runs/1014389285?check_suite_focus=true) fascinating.

- macos stable - 17m
- macos beta - 14m
- macos nightly - 51m

The speed of first two means we got assigned to the _fast_ macOS runners (the newer, slower ones take 20+ minutes).

The slowness of the last one is the sporadic macOS CI bug that hasn't been eliminated from the shared workers.  AFAICT the bug affects both the fast and slow runners.

- ubuntu stable - 22m
- ubuntu beta - 18m
- ubuntu nightly - 18m

In my initial testing, there were only the faster mac runners, so I was comparing the ~18m Linux runtime against ~14m macOS runs.  Given the current situation, Linux is a better choice.

File: .github/workflows/ci.yml
Comment (Veetaha) [reply .github/workflows/ci.yml:72]: As for me a simple `cargo foo` invocation looks descriptive enough by itself...

File: .github/workflows/ci.yml
Comment (Veetaha) [reply .github/workflows/ci.yml]: Done https://github.com/amethyst/amethyst/pull/2450/commits/213d343f7450e9e936248cf9fe03af89df9590f0

File: .github/workflows/ci.yml
Comment (Veetaha) [reply .github/workflows/ci.yml]: Done https://github.com/amethyst/amethyst/pull/2450/commits/213d343f7450e9e936248cf9fe03af89df9590f0

**Comment by Veetaha** [comment]
I've verified that linkcheck works in this commit https://github.com/amethyst/amethyst/pull/2450/commits/213d343f7450e9e936248cf9fe03af89df9590f0 in this ci run: https://github.com/amethyst/amethyst/pull/2450/checks?check_run_id=1028251375#step:13:11

**Comment by Blisto91** [comment]
I'm currently working on fixing all the broken links and linkcheck warnings.
Will also prepare the options in ``book.toml`` 👍

**Review by CleanCut — APPROVED** [review]

**Comment by Veetaha** [comment]
Whoops, what could go wrong...

**Comment by Veetaha** [comment]
Hmm, I can't reproduce the clippy warning locally...

**Comment by Blisto91** [comment]
Looks like it started with the newly released rust version.
But I didn't see it mentioned in the clippy changelog

**Comment by Veetaha** [comment]
This could be some bug fix in 1.46
I guess it might make sense to pin the stable version, run clippy at least for the beta toolchain or don't deny warnings on ci

**Comment by Blisto91** [comment]
I'm not a fan of preventing the CI from updating to the latest stable as soon as it's released. If thats what you mean by pin the stable version.
A minor semver version of rust is only released approx. every 6 weeks so i don't think it's too much of a bother if something breaks.
If we also enable clippy lints on beta we might risk that things break more often when bugs are introduced or false positivs comes up for a short time.

**Comment by Veetaha** [comment]
> I'm not a fan of preventing the CI from updating to the latest stable as soon as it's released. If thats what you mean by pin the stable version.
A minor semver version of rust is only released approx. every 6 weeks so i don't think it's too much of a bother if something breaks.
If we also enable clippy lints on beta we might risk that things break more often when bugs are introduced or false positivs comes up for a short time.

Since `clippy` may introduce new warnings to code that was fine on the previous minor toolchain version and 
taking this into account the @Blisto91 feedback I see these options:
- We accept that each 6 weeks our ci maybe failing due to new clippy lints and fix them ASAP
 This means unrelated PRs will be blocked until the fix.
- We make `clippy` a separate job and make it optional to pass via [`continue-on-error`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error). This will protect us from sudden breakages, though also allow merging the branch with clippy warnings, but the job should be displayed with a red cross so it's much easier to notice that clippy is failing
- We do remove `--deny warnings`. This means the warnings will most likely be ignored if some person doesn't regularly keep an eye on them.

Regarding the beta/nightly runs
I would also make them optional via [`continue-on-error`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontinue-on-error). In my experience we had a number of times the bugs were introduced to nightly toolchain so that our nightly job was failing. The purpose of running them is to verify our code forwards compatibility with thee toolchain updates and to give rust core team some feedback in case of regressions too.

**Comment by Blisto91** [comment]
The resistance we had to ``continue-on-error`` originally was because we didn't like how it was presented in the ui with the red crosses.
We were looking [for alternatives](https://github.com/actions/runner/issues/2347) but there aren't any that are properly supported by github.
But I can see that there are scenarios where it can make sense to use it.

I think the short to mid term aim should be to make life easier for the project with minimal manual maintenance (without compromising too much on quality). Currently the activity and reaction pace can be at bit slow so I would like to minimize scenarios where something has to be fixed quickly to rid errors popping up everywhere.
In the clippy case I'm personally leaning towards removing ``--deny warnings``

**Comment by Veetaha** [comment]
@CleanCut , @Blisto91 I think this is ready for `bors r+` to retry

**Pull request merged**
